File: Week_02/49.字母异位词分组.py
#
# @lc app=leetcode.cn id=49 lang=python3
#
# [49] 字母异位词分组
#

# @lc code=start
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        maps = dict()
        for item in strs:
            key = tuple(sorted(item))
            maps[key] = maps.get(key, []) + [item]
        return list(maps.values())
# @lc code=end

File: Week_09/49.字母异位词分组.cpp
/*
 * @lc app=leetcode.cn id=49 lang=cpp
 *
 * [49] 字母异位词分组
 */

// @lc code=start
class Solution
{
public:
    vector<vector<string>> groupAnagrams(vector<string> &strs)
    {
        vector<vector<string>> res;
        int sub = 0;
        string tmp;
        unordered_map<string, int> hashmap;
        for (const auto &str : strs)
        {
            tmp = str;
            sort(tmp.begin(), tmp.end());
            if (hashmap.count(tmp))
            {
                res[hashmap[tmp]].push_back(str);
            }
            else
            {
                vector<string> vec(1, str);
                res.push_back(vec);
                hashmap[tmp] = sub++;
            }
        }
        return res;
    }
};
// @lc code=end
